nixos/modules/services/web-apps/languagetool.nix

```diff
@@ -0,0 +1,73 @@
+{ config, pkgs, lib, ... }:
+with lib;
+let
+  cfg = config.services.languageToolHttp;
+  format = pkgs.formats.toml { };
+  confDoc = "See languagetool-http-server --help for further details";
+in
+{
+  options.services.languageToolHttp = {
+    enable = mkEnableOption "the LanguagTool http server";
+    allowBrowserPluginAccess = mkEnableOption "access from the browser plugin";
+    public = mkEnableOption ''
+      access this server from anywhere, not only localhost.
+      Note that it is reccomended to run this behind
+      a HTTPS reverse proxy like apache or ngnix
+    '';
+    settings = mkOption {
+      description = ''
+        Contents of the configuration file. ${confDoc}
+      '';
+      inherit (format) type;
+      default = { };
+    };
+    args = mkOption {
+      description = ''
+        Arguments to be passed when starting the server. ${confDoc}
+      '';
+      type = types.submodule {
+        freeformType = with types;attrsOf (oneOf [ int str path ]);
+        options.port = mkOption {
+          description = ''
+            The port which the LanguageTool server listen.
+          '';
+          type = types.port;
+          default = 8081;
+        };
+      };
+      default = { };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.languageToolHttp.args = {
+      allow-origin = mkIf cfg.allowBrowserPluginAccess (mkDefault "*");
+      config = format.generate "LanguageTool.cfg" cfg.settings;
+      public = mkIf cfg.public "";
+    };
+    systemd.services.languagetool-http = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      description = "LanguageTool HTTP server";
+      serviceConfig = {
+        User = "langtool";
+        Group = "langtool";
+        Type = "simple";
+        ExecStart = "${pkgs.languagetool}/bin/languagetool-http-server ${cli.toGNUCommandLineShell {} cfg.args}";
+        PrivateTmp = true;
+        NoNewPrivileges = true;
+        PrivateDevices = true;
+        PrivateMounts = true;
```

Comment:
Due to the hardening, it is not possible to just put a path into `settings` for e.g. pre-downloaded `languageModel`.

A solution should be to add a `StateDirectory` or to use `DynamicUser` (i.e. a less confined environment) to have a place for/access to these models.
`ReadOnlyPaths` **didn't** work for me with this config (Resulting in a `directory not found`):
```nix

  services.languageToolHttp = {
    enable = true;

    args = {
      port = 8081;
    };    
    
    settings = {
      fasttextBinary = "${pkgs.fasttext}/bin/fasttext";
      fasttextModel = "/var/lib/languagetool/lid.176.bin";

      # Approx. 10GB.
      # Data from:
      # https://languagetool.org/download/ngram-data/
      languageModel = "/var/lib/languagetool/languageModel";
      
      # Download all data here! Otherwise, it fails to start
      # Data from:
      # https://languagetool.org/download/word2vec/
      word2vecModel = "/var/lib/languagetool/word2vec";
      
      pipelinePrewarming = true;
      prometheusMonitoring = true;
      prometheusPort = 9301;
    };
  };

  systemd.services.languagetool-http.serviceConfig = {
    ReadOnlyPaths = [
      "/var/lib/languagetool/"
    ];
  };
  ```

Reply:
I would ask you, as I do not use that functionality, what would be a good interface.

In the interim, have you tried [BindReadOnlyPaths=](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#BindPaths=)?

Reply:
`BindReadOnlyPaths=` did not work for me, with the same result:
```
Exception in thread "main" java.lang.RuntimeException: LanguageModel directory not found or is not a directory: "/var/lib/languagetool/languageModel"
```

But I now noticed the problem: TOML is the wrong format, the configuration is actually a Java properties file, similar to what [minecraft `server.properties`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/games/minecraft-server.nix#L18-L23) does. Comparing the outputs configurations, the main difference is the **absence of quotations**. So obviously, the path `"/var/lib/...` will not exist, because it tries to open the (relative) directory `./"/var/lib/...`.

With this change in mind, `BindReadOnlyPaths` works - actually, no additional configuration is needed at all with the current state of this PR, i.e. with `DynamicUser=true`.